From cc42960d8619c1fba4e1057070aab7cc28a48948 Mon Sep 17 00:00:00 2001
From: Russell Coker <russell@coker.com.au>
Date: Tue, 18 Aug 2026 23:46:17 +1000
Subject: udev: own type for udev tmpfs files (memfd)

Backport of the udev.te part of refpolicy cc42960 ("tmpfs (#1200)").
The refpolicy this layer builds on (cffa6e2c, meta-selinux 1ba26da)
predates it.

Without it, the memfd systemd-udevd creates on the kernel's internal
tmpfs is labelled tmpfs_t and udev_t is denied getattr on it. Observed
on Rock 5A in persistent journals from several boots, the current image
included:

  avc: denied { getattr } comm="systemd-udevd"
  path="/memfd:systemd-udevd (deleted)" dev="tmpfs"
  scontext=system_u:system_r:udev_t:s0
  tcontext=system_u:object_r:tmpfs_t:s0 tclass=file permissive=0

With the type transition the memfd gets udev_tmpfs_t, which udev_t
manages. Drop this patch when the layer moves to a refpolicy that
contains cc42960.

Upstream-Status: Backport [https://github.com/SELinuxProject/refpolicy/commit/cc42960d8619c1fba4e1057070aab7cc28a48948]
---
diff --git a/policy/modules/system/udev.te b/policy/modules/system/udev.te
index 81fe0606..0b76f7b2 100644
--- a/policy/modules/system/udev.te
+++ b/policy/modules/system/udev.te
@@ -33,6 +33,9 @@ optional_policy(`
 	kubernetes_mountpoint(udev_runtime_t)
 ')
 
+type udev_tmpfs_t;
+files_tmpfs_file(udev_tmpfs_t)
+
 ########################################
 #
 # udev Local policy
@@ -87,6 +90,9 @@ manage_sock_files_pattern(udev_t, udev_runtime_t, udev_runtime_t)
 files_runtime_filetrans(udev_t, udev_runtime_t, dir, "udev")
 allow udev_t udev_runtime_t:dir watch;
 
+allow udev_t udev_tmpfs_t:file manage_file_perms;
+fs_tmpfs_filetrans(udev_t, udev_tmpfs_t, file)
+
 kernel_load_module(udev_t)
 kernel_read_system_state(udev_t)
 kernel_request_load_module(udev_t)
